opm/simulators/linalg/bda/WellContributionsOCLContainer.cpp

```diff
@@ -127,13 +135,50 @@ namespace bda
                                  cl::Local(lmem1), cl::Local(lmem2), cl::Local(lmem2));
     }
 
+
+    void WellContributionsOCLContainer::applyMSWells(cl::Buffer& x, cl::Buffer& y) {
+        cl::Event event;
+        std::vector<cl::Event> events(2);
+
+        // copy vectors x and y from GPU to CPU
+        queue->enqueueReadBuffer(x, CL_FALSE, 0, sizeof(double) * N, x_msw.data(), nullptr, &events[0]);
+        queue->enqueueReadBuffer(y, CL_FALSE, 0, sizeof(double) * N, y_msw.data(), nullptr, &events[1]);
+        event.waitForEvents(events);
+
+        // actually apply MultisegmentWells
+        for(Opm::MultisegmentWellContribution *well: multisegments){
+            well->setReordering(toOrder.data(), true);
+            well->apply(x_msw.data(), y_msw.data());
+        }
+
+        // copy vector y from CPU to GPU
+        queue->enqueueWriteBuffer(y, CL_FALSE, 0, sizeof(double) * N, y_msw.data(), nullptr, &event);
+        event.wait();
+    }
+
     void WellContributionsOCLContainer::apply(cl::Buffer& x, cl::Buffer& y){
         if(num_std_wells > 0){
             applyStdWells(x, y);
         }
+
+        if(num_ms_wells > 0){
+            applyMSWells(x, y);
+        }
     }
 
     WellContributionsOCLContainer::~WellContributionsOCLContainer(){
-        toOrder.clear();
+        if(num_std_wells > 0){
+            toOrder.clear();
```

Comment:
I don't think you have to explicitly clear the containers. std::vector will do that in the destructor anyway

Reply:
Exactly. The only thing this constructor needs to do is to delete the new-ed pointers in the `multisegments` vector. However, the class is missing copy constructor, move constructor, copy assignment operator and move assignment operator. This is the "rule of 5" (used to be the "rule of 3" before move operations came in C++11), if you must implement one, you must implement all of them. As is now, copying this object will cause a crash, as the mentioned pointers will be double-deleted. A better approach is to use smart pointers. If you use unique_ptr instead, you do not have to implement any of these functions, but can safely use the defaults. The copy operations will then not be available since unique_ptr can be moved, not copied.

Reply:
Simply delete the copy constructor then:

```
WellContributionsOCLContainer
{
  WellContributionsOCLContainer(const WellContributionsOCLContainer&) = delete;
}
```

Reply:
I tried to add the smart pointers for the move operations in the `WellContributionsOCLContainer` class, however I'm not sure If what I did is what you guys recommended. I still have some doubts of where the destructor for the components of the multisegment wells vector should be (I've put it in the `WellContributionsOCLContainer` class since that is where the vector is moved to). I didn't write the copy constructor, I guess I didn't really understand what that is for.

Anyways, let me know what needs to be fixed.

Reply:
Seasoned programmers should not try to impress others in reviews. This time this could just have wasted your time (and mine, too). I will take a look.

Reply:
> Seasoned programmers should not try to impress others in reviews.

In all seriousness, I just tried to be helpful. I think your suggest to delete the copy constructor is fine as well (but should also delete the copy assignment operator). I do not think it is wasteful to point out bad patterns in code, not sure what ticked you off here?

Reply:
It is not 100% clear was @atgeirr was saying or trying to say. I think he should clarify things and maybe should have taken a closer look at the whole code.

Maybe he meant, to use a smartpointer for each entry of the container (and not the container itself)? Then no destructor would be needed and you would not need to add more methods. The problem is that unique_ptr is only movable and not copyable. Hence you cannot use all the functionality of std::vector (e.g. resizing, etc.) and you will have to use push or emplace entries in a special way with `std::move`). That seems like quite a burden to put on people new to the language, and will come as
a surprise to others if they are trying to use std::vector the usual way. Maybe he meant what you did use a std::unique_ptr<std::vector> >?

Otherwise in principle one should implement all five mentioned De- and Constructors to prevent surpises and make the code correct for every use case  is of course very good and should be adhered to. But this is still an internal class and more widespread usage is not foreseeable, yet. I guess there are quite some places in OPM where this is also not done yet, but should, with higher priority.

Indeed the problem that pointers might be deleted twice is gone now (as one cannot copy the class anymore), but the code complexity has increased quite a bit for that.

I am tempted to pass the GPU review torch to @atgeirr.

Reply:
@atgeirr Can you please take a look at the changes you triggered and see whether that is what you suggested?

We now have a `std::unique_ptr<std::vector<MultiSegementWellContribution*>> multisegments` instead of a simple vector and that is initialized by copying ` multisegments = std::move(std::make_unique<mswVecT>(wellContribs.multisegments));`. Destructor is basically unchanged.